### PR TITLE
Disallow 'var' as an explicit lambda expression return type

### PR DESCRIPTION
**Pull request by cston**
See [LDM-2021-06-02 notes](https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-06-02.md#lambda-return-type-parsing).

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -12764,15 +12764,18 @@ LambdaExpressionSyntax parseLambdaExpressionWorker()
...
                 }
 
-                TypeSyntax returnType;
+                TypeSyntax returnType = null;
                 var resetPoint = this.GetResetPoint();
                 try
                 {
-                    returnType = ParseReturnType();
-                    if (CurrentToken.Kind != SyntaxKind.OpenParenToken)
+                    if (!IsVarType())
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:12771]: >if (!IsVarType()) [](http://example.com/codeflow?start=20&length=17)

Should we instead parse this as a type and report an error that "var" is not allowed as a return type of a lambda?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:12771]: Perhaps even in binding.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -12787,6 +12787,12 @@ LambdaExpressionSyntax parseLambdaExpressionWorker()
                     arrow = CheckFeatureAvailability(arrow, MessageID.IDS_FeatureLambda);
                     var (block, expression) = ParseLambdaBody();
 
+                    // Disallow 'var' as explicit return type.
+                    if ((returnType as IdentifierNameSyntax)?.Identifier.ContextualKind == SyntaxKind.VarKeyword)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:12791]: >if ((returnType as IdentifierNameSyntax)?.Identifier.ContextualKind == SyntaxKind.VarKeyword) [](http://example.com/codeflow?start=20&length=93)

Should we also disallow ```ref var```?

File: src/Compilers/CSharp/Test/Syntax/Parsing/LambdaReturnTypeParsingTests.cs [hunk 3]
```diff
@@ -5068,6 +5077,229 @@ public void Dynamic_02()
...
+        public void Var_05()
+        {
+            string source = "var d = var () => default;";
+            verify(source, TestOptions.Regular9);
+            verify(source);
+
+            void verify(string source, ParseOptions? parseOptions = null)
+            {
+                UsingDeclaration(source, parseOptions,
+                    // (1,9): error CS1031: Type expected
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LambdaReturnTypeParsingTests.cs:5259]: >error CS1031: Type expected [](http://example.com/codeflow?start=30&length=27)

It feels like this generic error might confuse a user.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Syntax/Parsing/LambdaReturnTypeParsingTests.cs:5259]: Perhaps `error CS0825: The contextual keyword 'var' may only appear within a local variable declaration or in script code` is suitable. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQGYACLEgYRIG8CT6SA3CAJxIA8BuOh4sgFhIBZABQBKGj3oBfAtKA===)

**Comment by jcouv** [comment]
Moved PR to target `main`/17.1

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs [hunk 4]
```diff
@@ -3277,21 +3278,47 @@ class Program
...
             CompileAndVerify(source, parseOptions: TestOptions.Regular9, expectedOutput:
-@"System.Func`1[System.Func`1[System.Object]]");
+@"System.Func`1[System.Func`1[System.Object]]
+System.Func`1[System.Func`1[System.Object]]
+System.Func`1[System.Func`1[System.Object]]
+");
 
             // Breaking change from C#9 which binds calls to F(Func<Func<object>>, int).
+            //
+            // The calls such as F(() => () => 1, 2) should be considered ambiguous in C#9 as per the C# spec.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:3301]: >// The calls such as F(() => () => 1, 2) should be considered ambiguous in C#9 as per the C# spec. [](http://example.com/codeflow?start=12&length=98)

The change doesn't look related to the goal of the PR. Is it intentional?

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs [hunk 5]
```diff
@@ -248,6 +248,11 @@ static void checkAttributes(AnonymousFunctionExpressionSyntax syntax, SyntaxList
             MessageID.IDS_FeatureLambdaReturnType.CheckFeatureAvailability(diagnostics, syntax);
 
             syntax = syntax.SkipRef(out RefKind refKind);
+            if ((syntax as IdentifierNameSyntax)?.Identifier.ContextualKind() == SyntaxKind.VarKeyword)
+            {
+                diagnostics.Add(ErrorCode.ERR_LambdaExplicitReturnTypeVar, syntax.Location);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:253]: Probably should document as breaking change in 17.1

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:253]: Does the spec capture that design decision?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7) modulo documentation questions

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs [hunk 6]
```diff
@@ -5369,6 +5369,199 @@ static void Main()
...
+        public void VarReturnType_05()
+        {
+            var source =
+@"using System;
+using var = System.Int32;
+class Program
+{
+    static void Main()
+    {
+        F(var (var v) => v);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:5503]: >var [](http://example.com/codeflow?start=10&length=3)

Why this `var` is not reported?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs:5503]: Perhaps there is a mismatch in source for VerifyDiagnostics below.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs:253]: I've added an item to the breaking changes document. I'll update the spec in a subsequent PR.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 9)
